[US2703] support list snapshots capability in zrepl

mynktl opened the pull request:
Changes : 
Adding the handling of snapshot list command in ZREPL.
A response message from ZREPL will contain the snapshot list and properties in JSON format as described below : 
```json
{
    "snapshot”: [ {
        {
            “name” : “snapshot1 name”,
            “properties”  : { 
                “prop1”:”val1”,
                “prop2”: “val2”,
            },
        },
        {
            “name” : “snapshot2 name”,
            “properties”  : { 
                “prop1”:”val1”,
                “prop2”: “val2”,
                “prop3” : “val3”
            },
        },]
}
```
This is an initial version of changes. By default, we are sending all the properties of the snapshot to the controller.

satbirchhikara committed:
[DE81] Do not free zio_cmd in uzfs_zvol_worker if zio_cmd is meant for rebuild. (#92)

Signed-off-by: satbir <satbir.chhikara@gmail.com>

Commit:
[DE83] update zvol status as degraded when target connection breaks

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta committed:
[DE84] Reconnecting to disconnected targets periodically, and, setting keepalive parameters for connections (#95)

* Setting keepalive on connected sockets
* scan for any pending connects to target atleast every 2 seconds

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

vishnuitta committed:
[DE83] Reinitializing zv variables during data connection break (#94)

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

Commit:
Adding support for snapshot list command in zrepl

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta committed:
[DE85] Rebuild thread need to be exited when sender thread exits (#96)

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

satbirchhikara committed:
[TA2330] API to store snapshot IO num when taking snapshot, to retrieve snapshot dataset and IO number from given snapname (#98)

Signed-off-by: satbir <satbir.chhikara@gmail.com>

vishnuitta committed:
Data connection creation issue during zrepl restart (#99)

* testcase added, and, added flag to track io receiver
Signed-off-by: Vishnu Itta <vitta@mayadata.io>

vishnuitta committed:
lock issue in error case (#100)

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

mynktl committed:
[TA2577] Updating ZAP (degraded_io_seq) at every 5 seconds if zvol is degraded (#101)

- Initializing zinfo's checkpointed io sequence number with old value stored in ZAP

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

jkryl commented on lib/libzrepl/mgmt_conn.c:
```diff
@@ -1017,6 +1159,32 @@ process_message(uzfs_mgmt_conn_t *conn)
+			    hdrp->opcode, hdrp->io_seq);
+			break;
+		}
+		strlcpy(zvol_name, payload, payload_size);
```
payload_size -> sizeof (zvol_name)

jkryl commented on lib/libzrepl/mgmt_conn.c:
```diff
@@ -555,6 +558,145 @@ uzfs_zvol_stats(uzfs_mgmt_conn_t *conn, zvol_io_hdr_t *hdrp, zvol_info_t *zinfo)
+	hdr.status = ZVOL_OP_STATUS_OK;
+
+	return (reply_data(conn, &hdr, snap_list,
+	    total_len + sizeof (*snap_list)));
```
if i understand correctly this is called on behalf of asynchronous thread dispatched from mgmt thread. You can't touch tcp connection or send anything on it from this context. Only the mgmt thread can do such things. So once you complete the list of snapshots, you should return this result back to the mgmt thread and call reply_data() there. Just like we do with other async operations.

mynktl committed:
- Addressing review comments from PR#101 (#102)

- Change in LOG message if CONFIG_LOAD_ENABLE is set

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

pawanpraka1 committed:
fixing nvlist leak during IOCTL handling (#103)

Signed-off-by: Pawan <pawanprakash101@gmail.com>

vishnuitta committed:
ARC informed only 1/4 of phys_mem as all_mem

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

vishnuitta committed:
Adding signal handler for FPE/ABRT

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

satbirchhikara committed:
[DE95]Throttle rebuild in replica which helps in rebuild so that container do not cross memory limits. (#106)

Signed-off-by: satbir <satbir.chhikara@gmail.com>

vishnuitta committed:
[DE94]fix(zfs):zrepl crash during data conn break (#109)

When data connection breaks, volume's metavolblocksize is made to 0.
If any IOs are pending during this time, executing them will use this
and causes floating point exception.

Fix is NOT to make metavolblocksize as 0, as this is not required.

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

satbirchhikara committed:
[TA2335]Create snapshot and clone for rebuild, the moment data connection is created. (#108)

Signed-off-by: satbir <satbir.chhikara@gmail.com>

vishnuitta commented on lib/libzrepl/mgmt_conn.c:
```diff
@@ -1017,6 +1159,32 @@ process_message(uzfs_mgmt_conn_t *conn)
+			break;
+		}
+		LOGCONN(conn, "Snaplist command for %s", zinfo->name);
+		rc = uzfs_zvol_fetch_snapshot_list(conn, hdrp, zinfo);
```
this need to be called from async thread, as, we have a single mgmt_thread for all volumes. Doing this call here can block mgmt_thread for other volumes as well.

vishnuitta replied:
please look at SNAPCREATE opcode

vishnuitta commented on lib/libzrepl/mgmt_conn.c:
```diff
@@ -1017,6 +1159,32 @@ process_message(uzfs_mgmt_conn_t *conn)
 		break;
 
+	case ZVOL_OPCODE_SNAP_LIST:
+		if (payload_size == 0 || payload_size > MAX_NAME_LEN) {
```
this code seems to be repetition of SNAPCREATE opcode. Need to remove duplicate code.

vishnuitta committed:
feat(zfs): store snapshot IO number in ZAP as healthy_ioseq (#110)

Signed-off-by: Vishnu Itta <vitta@mayadata.io>

Commit:
- Addressing review comments
- Merge remote-tracking branch 'upstream/zfs-0.7-release' into HEAD

Signed-off-by: mayank <mayank.patel@cloudbyte.com>
